File: excel_restaurant_pos/excel_restaurant_pos/report/performance_report/performance_report.py
# Copyright (c) 2025, Sohanur Rahman and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from datetime import datetime

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {
            "label": _("Order ID"),
            "fieldname": "order_id",
            "fieldtype": "Link",
            "options": "Table Order",
            "width": 120
        },
        # {
        #     "label": _("Customer"),
        #     "fieldname": "customer_name",
        #     "fieldtype": "Data",
        #     "width": 150
        # },
        {
            "label": _("Item"),
            "fieldname": "item",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": _("Qty"),
            "fieldname": "qty",
            "fieldtype": "Float",
            "width": 80
        },
        {
            "label": _("Order Placed"),
            "fieldname": "order_placed_time_formatted",
            "fieldtype": "Data",
            "width": 180
        },
        # {
        #     "label": _("Order Ready"),
        #     "fieldname": "order_ready_time",
        #     "fieldtype": "Datetime",
        #     "width": 150
        # },
        # {
        #     "label": _("Order Confirmed"),
        #     "fieldname": "order_confirm_time",
        #     "fieldtype": "Datetime",
        #     "width": 150
        # },
        {
            "label": _("Placed to Accepted (mins)"),
            "fieldname": "placed_to_accepted_duration",
            "fieldtype": "Float",
            "width": 200,
            "precision": 2
        },
        {
            "label": _("Accepted to Ready (mins)"),
            "fieldname": "accepted_to_ready_duration",
            "fieldtype": "Float",
            "width": 200,
            "precision": 2
        },
        {
            "label": _("Ready to Confirmed (mins)"),
            "fieldname": "ready_to_confirmed_duration",
            "fieldtype": "Float",
            "width": 150,
            "precision": 2
        },
        {
            "label": _("Total Duration (mins)"),
            "fieldname": "total_duration",
            "fieldtype": "Float",
            "width": 200,
            "precision": 2
        },
        {
            "label": _("Status"),
            "fieldname": "status",
            "fieldtype": "Data",
            "width": 100
        }
    ]

def get_data(filters=None):
    conditions = get_conditions(filters)
    
    query = """
        SELECT 
            toi.parent as order_id,
            to_order.customer_name,
            toi.item,
            toi.qty,
            toi.order_placed_time,
            toi.order_accepted_time,
            toi.order_ready_time,
            toi.order_confirm_time,
            toi.is_accepted,
            toi.is_ready,
            to_order.status,
            to_order.creation as order_creation
        FROM `tabTable Order Item` as toi
        LEFT JOIN `tabTable Order` as to_order ON toi.parent = to_order.name
        WHERE to_order.docstatus = 1 {conditions}
        ORDER BY toi.parent DESC, toi.idx ASC
    """.format(conditions=conditions)
    
    data = frappe.db.sql(query, as_dict=1)
    
    # Calculate durations and format dates
    for row in data:
        # Format order_accepted_time with proper error handling
        if row.order_placed_time:
            try:
                if isinstance(row.order_placed_time, str):
                    accepted_datetime = datetime.strptime(row.order_placed_time, '%Y-%m-%d %H:%M:%S.%f')
                else:
                    accepted_datetime = row.order_placed_time
                
                # Format as: 10-20-2023 04:56pm
                row['order_placed_time_formatted'] = accepted_datetime.strftime('%m-%d-%Y %I:%M%p').lower()
            except Exception:
                # Fallback to original value if formatting fails
                row['order_placed_time_formatted'] = str(row.order_placed_time) if row.order_accepted_time else None
        else:
            row['order_placed_time_formatted'] = None
        
        # Calculate duration between stages
        placed_to_accepted = calculate_duration(row.order_placed_time, row.order_accepted_time)
        accepted_to_ready = calculate_duration(row.order_accepted_time, row.order_ready_time)
        ready_to_confirmed = calculate_duration(row.order_ready_time, row.order_confirm_time)
        
        # Total duration from placed to confirmed
        total_duration = calculate_duration(row.order_placed_time, row.order_confirm_time)
        
        row.update({
            'placed_to_accepted_duration': placed_to_accepted,
            'accepted_to_ready_duration': accepted_to_ready,
            'ready_to_confirmed_duration': ready_to_confirmed,
            'total_duration': total_duration
        })
    
    return data

def calculate_duration(start_time, end_time):
    """Calculate duration between two datetime objects in minutes"""
    if not start_time or not end_time:
        return None
    
    try:
        if isinstance(start_time, str):
            start_time = datetime.strptime(start_time, '%Y-%m-%d %H:%M:%S.%f')
        if isinstance(end_time, str):
            end_time = datetime.strptime(end_time, '%Y-%m-%d %H:%M:%S.%f')
        
        duration = end_time - start_time
        return round(duration.total_seconds() / 60, 2)  # Convert to minutes
    except:
        return None

def get_conditions(filters):
    conditions = ""
    
    if filters:
        if filters.get("from_date"):
            conditions += " AND DATE(to_order.creation) >= %(from_date)s"
        
        if filters.get("to_date"):
            conditions += " AND DATE(to_order.creation) <= %(to_date)s"
        
        if filters.get("customer"):
            conditions += " AND to_order.customer = %(customer)s"
        
        if filters.get("item"):
            conditions += " AND toi.item = %(item)s"
        
        if filters.get("status"):
            conditions += " AND to_order.status = %(status)s"
    
    return conditions

# Additional utility functions for summary statistics
def get_summary_data(filters=None):
    """Get summary statistics for performance metrics"""
    conditions = get_conditions(filters)
    
    summary_query = """
        SELECT 
            COUNT(*) as total_orders,
            AVG(TIMESTAMPDIFF(MINUTE, toi.order_placed_time, toi.order_accepted_time)) as avg_placed_to_accepted,
            AVG(TIMESTAMPDIFF(MINUTE, toi.order_accepted_time, toi.order_ready_time)) as avg_accepted_to_ready,
            AVG(TIMESTAMPDIFF(MINUTE, toi.order_ready_time, toi.order_confirm_time)) as avg_ready_to_confirmed,
            AVG(TIMESTAMPDIFF(MINUTE, toi.order_placed_time, toi.order_confirm_time)) as avg_total_duration,
            MAX(TIMESTAMPDIFF(MINUTE, toi.order_placed_time, toi.order_confirm_time)) as max_total_duration,
            MIN(TIMESTAMPDIFF(MINUTE, toi.order_placed_time, toi.order_confirm_time)) as min_total_duration
        FROM `tabTable Order Item` as toi
        LEFT JOIN `tabTable Order` as to_order ON toi.parent = to_order.name
        WHERE to_order.docstatus = 1 
        AND toi.order_placed_time IS NOT NULL 
        AND toi.order_confirm_time IS NOT NULL
        {conditions}
    """.format(conditions=conditions)
    
    return frappe.db.sql(summary_query, as_dict=1)

File: excel_restaurant_pos/excel_restaurant_pos/report/performance_report/performance_report.js
// Copyright (c) 2025, Sohanur Rahman and contributors
// For license information, please see license.txt

frappe.query_reports["Table Order Performance Report"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "width": "80"
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "width": "80"
        },
        {
            "fieldname": "customer",
            "label": __("Customer"),
            "fieldtype": "Link",
            "options": "Customer",
            "width": "100"
        },
        {
            "fieldname": "item",
            "label": __("Item"),
            "fieldtype": "Link",
            "options": "Item",
            "width": "100"
        },
        {
            "fieldname": "status",
            "label": __("Order Status"),
            "fieldtype": "Select",
            "options": ["", "Pending", "In Progress", "Completed", "Cancelled"],
            "width": "100"
        }
    ],
    
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        // Highlight slow orders in red
        if (column.fieldname === "total_duration" && data && data.total_duration > 30) {
            value = `<span style='color:red; font-weight: bold;'>${value}</span>`;
        }
        
        // Highlight fast orders in green
        else if (column.fieldname === "total_duration" && data && data.total_duration < 10) {
            value = `<span style='color:green; font-weight: bold;'>${value}</span>`;
        }
        
        // Color code durations
        if (column.fieldname.includes("_duration") && data) {
            let duration_value = data[column.fieldname];
            if (duration_value > 15) {
                value = `<span style='color:orange;'>${value}</span>`;
            } else if (duration_value > 30) {
                value = `<span style='color:red;'>${value}</span>`;
            }
        }
        
        return value;
    },
    
    "onload": function(report) {
        // Add a summary button
        report.page.add_inner_button(__("Performance Summary"), function() {
            frappe.call({
                method: "frappe.desk.query_report.run",
                args: {
                    report_name: "Table Order Performance Report",
                    filters: report.get_values()
                },
                callback: function(r) {
                    if (r.message && r.message.result) {
                        show_performance_summary(r.message.result);
                    }
                }
            });
        });
    }
};

function show_performance_summary(data) {
    if (!data || data.length === 0) {
        frappe.msgprint(__("No data available for summary"));
        return;
    }
    
    let total_orders = data.length;
    let completed_orders = data.filter(row => row.order_confirm_time).length;
    
    // Calculate averages
    let total_durations = data.filter(row => row.total_duration).map(row => row.total_duration);
    let avg_total_duration = total_durations.length > 0 ? 
        (total_durations.reduce((a, b) => a + b, 0) / total_durations.length).toFixed(2) : 0;
    
    let placed_to_accepted = data.filter(row => row.placed_to_accepted_duration)
        .map(row => row.placed_to_accepted_duration);
    let avg_placed_to_accepted = placed_to_accepted.length > 0 ? 
        (placed_to_accepted.reduce((a, b) => a + b, 0) / placed_to_accepted.length).toFixed(2) : 0;
    
    let accepted_to_ready = data.filter(row => row.accepted_to_ready_duration)
        .map(row => row.accepted_to_ready_duration);
    let avg_accepted_to_ready = accepted_to_ready.length > 0 ? 
        (accepted_to_ready.reduce((a, b) => a + b, 0) / accepted_to_ready.length).toFixed(2) : 0;
    
    let ready_to_confirmed = data.filter(row => row.ready_to_confirmed_duration)
        .map(row => row.ready_to_confirmed_duration);
    let avg_ready_to_confirmed = ready_to_confirmed.length > 0 ? 
        (ready_to_confirmed.reduce((a, b) => a + b, 0) / ready_to_confirmed.length).toFixed(2) : 0;
    
    let summary_html = `
        <div class="performance-summary">
            <h4>Performance Summary</h4>
            <table class="table table-bordered">
                <tr><td><b>Total Orders:</b></td><td>${total_orders}</td></tr>
                <tr><td><b>Completed Orders:</b></td><td>${completed_orders}</td></tr>
                <tr><td><b>Completion Rate:</b></td><td>${((completed_orders/total_orders)*100).toFixed(1)}%</td></tr>
                <tr><td><b>Average Total Duration:</b></td><td>${avg_total_duration} minutes</td></tr>
                <tr><td><b>Average Placed to Accepted:</b></td><td>${avg_placed_to_accepted} minutes</td></tr>
                <tr><td><b>Average Accepted to Ready:</b></td><td>${avg_accepted_to_ready} minutes</td></tr>
                <tr><td><b>Average Ready to Confirmed:</b></td><td>${avg_ready_to_confirmed} minutes</td></tr>
            </table>
        </div>
    `;
    
    frappe.msgprint({
        title: __("Performance Summary"),
        message: summary_html,
        wide: true
    });
}